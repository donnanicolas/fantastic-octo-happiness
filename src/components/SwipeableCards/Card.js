// @flow
import React from 'react';
import styled from 'styled-components';
import { Thumbnail } from 'react-bootstrap';

import Preview from '../Preview';

const SizedThumbnail = styled(Thumbnail) `
  width: 240px;
  height: 378px;
`;

const ellipsis = (component: ReactClass<any>) => styled(component)`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const Header = ellipsis(({ children, ...props }) => <h3 {...props}>{children}</h3>);
const Paragraph = ellipsis(({ children, ...props }) => <p {...props}>{children}</p>);

type CardProps = {
  track: Track,
  style?: any,
  className?: string,
};

const Card = ({ track, style, className }: CardProps) => {
  const { image, album, name, artists } = track;
  const artistsString = artists
    .map(({ name }) => name)
    .join(', ');

  return (
    <SizedThumbnail src={image} alt={album} style={style} className={className}>
      <Header>{name}</Header>
      <Paragraph>{artistsString} - {album}</Paragraph>
      <div>
        <Preview track={track} />
      </div>
    </SizedThumbnail>
  );
};

Card.defaultProps = { style: {} };

const BackgroundCard = styled(Card) `
  position: relative;
  margin-top: -398px;
  z-index: -1;
`;

export default Card;
export {
  BackgroundCard,
};