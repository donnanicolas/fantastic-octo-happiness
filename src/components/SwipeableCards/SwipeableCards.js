// @flow
import React, { PureComponent } from 'react';
import Swipe from 'react-swipe-component';
import { Motion, spring, presets } from 'react-motion';
import styled from 'styled-components';

import Card, { BackgroundCard } from '../Card';

const { gentle } = presets;

type Props = {
  tracks: Track[],
  onSwipedLeft: (track: Track) => any,
  onSwipedRight: (track: Track) => any,
  onLoadMore: () => any,
};

type State = {
  tracks: Track[],
  processedTracks: Track[],
  animation: 'none' | 'swipeLeft' | 'swipeRight',
};

const PositionedSwipe = styled(Swipe)`
  position: relative;
  z-index: 2;
`;

const defaultCurrentStyle = { rotate: 0, translate: 0, opacity: 1 };
const leftSwipeStyle = { rotate: spring(-13, gentle), translate: spring(-100, gentle), opacity: spring(0, gentle) }
const rightSwipeStyle = { rotate: spring(13, gentle), translate: spring(100, gentle), opacity: spring(0, gentle) }

const defaultNextStyle = { scale: 0.75, opacity: 0 };
const scaleStyle = { scale: spring(1, gentle), opacity: 1 };

class SwipeableCard extends PureComponent {

  constructor(props: Props) {
    super(props);
    const { tracks } = props;
    this.state = {
      tracks,
      processedTracks: [],
      animation: 'none',
    }
  }

  static defaultProps: any;
  state: State;
  props: Props;

  swipeMotionStyle = () => {
    const { animation } = this.state;
    if (animation === 'swipeLeft') return leftSwipeStyle;
    if (animation === 'swipeRight') return rightSwipeStyle;
    return defaultCurrentStyle;
  }

  enterMotionStyle = () => {
    const { animation } = this.state;
    if (animation === 'none') return defaultNextStyle;
    return scaleStyle;
  }

  handleSwipeRight = (track: Track) => () => {
    this.setState(state => ({ animation: 'swipeRight' }));
    this.props.onSwipedRight(track);
  }

  handleSwipeLeft = (track: Track) => () => {
    this.setState(state => ({ animation: 'swipeLeft' }));
    this.props.onSwipedLeft(track);
  }

  handleOnRest = () => {
    this.setState(
      state => ({ animation: 'none' }),
      () => {
        this.setState(state => {
          const { tracks, processedTracks } = this.state;
          const [track, ...others] = tracks;
          const newProcessed = processedTracks.concat([track]);
          return { processedTracks: newProcessed, tracks: others };
        })
      });
  }

  render() {
    const { onLoadMore } = this.props;
    const { tracks } = this.state;
    const [track, nextTrack] = tracks;

    return (
      <div>
        <Motion
          defaultStyle={defaultCurrentStyle}
          style={this.swipeMotionStyle()}
        >
          {({ rotate, translate, opacity }) =>
            <PositionedSwipe
              nodeName="div"
              mouseSwipe
              delta={150}
              onSwipedLeft={this.handleSwipeLeft(track)}
              onSwipedRight={this.handleSwipeRight(track)}
              style={{ opacity, transform: `rotate(${rotate}deg) translate3d(${translate}%, 0, 0)` }}
            >
              <Card track={track} onLoadMore={onLoadMore} />
            </PositionedSwipe>
          }
        </Motion>
        <Motion
          defaultStyle={defaultNextStyle}
          style={this.enterMotionStyle()}
          onRest={this.handleOnRest}
        >
          {({ scale, opacity }) =>
            <BackgroundCard
              style={{ opacity, transform: `scale(${scale})` }}
              onLoadMore={onLoadMore}
              track={nextTrack}
            />
          }
        </Motion>
      </div>
    );
  }
}

SwipeableCard.defaultProps = {
  onSwipedLeft: () => { },
  onSwipedRight: () => { },
}

export default SwipeableCard;
