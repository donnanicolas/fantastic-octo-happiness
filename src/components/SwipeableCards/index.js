// @flow
import SwipeableCards from './SwipeableCards';
export default SwipeableCards;
