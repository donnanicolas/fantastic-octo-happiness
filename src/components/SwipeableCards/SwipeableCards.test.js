import React from 'react';
import SwipeableCards from './SwipeableCards';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';

import tracks from '../../fixtures/tracks';
 
const createClientXY = (x, y) => ({ clientX: x, clientY: y });
const createTouchEventObject = (x = 0, y = 0) => ({ touches: [createClientXY(x, y)] });

it('renders a card', () => {
  const tree = renderer.create(
    <div >
      <SwipeableCards
        tracks={tracks}
      />
    </div>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should call function on swipe left with the mouse', () => {
  const onSwipedLeft = sinon.spy();
  const wrapper = mount(
    <SwipeableCards
      tracks={tracks}
      onSwipedLeft={onSwipedLeft}
    />
  );

  const Card = wrapper.find('.Card');

  Card.simulate('mouseDown', createClientXY(250, 50));
  Card.simulate('mouseMove', createClientXY(50, 50));
  Card.simulate('mouseUp', createClientXY(50, 50));
  
  expect(onSwipedLeft).toHaveProperty('callCount', 1);
});

it('should call function on swipe right with the mouse', () => {
  const onSwipedRight = sinon.spy();
  const wrapper = mount(
    <SwipeableCards
      tracks={tracks}
      onSwipedRight={onSwipedRight}
    />
  );

  const Card = wrapper.find('.Card');

  Card.simulate('mouseDown', createClientXY(50, 50));
  Card.simulate('mouseMove', createClientXY(250, 50));
  Card.simulate('mouseUp', createClientXY(250, 50));
  
  expect(onSwipedRight).toHaveProperty('callCount', 1);
});

it('should call function on swipe left with the touch', () => {
  const onSwipedLeft = sinon.spy();
  const wrapper = mount(
    <SwipeableCards
      tracks={tracks}
      onSwipedLeft={onSwipedLeft}
    />
  );

  const Card = wrapper.find('.Card');

  Card.simulate('touchStart', createTouchEventObject(250, 50));
  Card.simulate('touchMove', createTouchEventObject(50, 50));
  Card.simulate('touchEnd', createTouchEventObject(50, 50));
  
  expect(onSwipedLeft).toHaveProperty('callCount', 1);
});

it('should call function on swipe left with the touch', () => {
  const onSwipedRight = sinon.spy();
  const wrapper = mount(
    <SwipeableCards
      tracks={tracks}
      onSwipedRight={onSwipedRight}
    />
  );

  const Card = wrapper.find('.Card');

  Card.simulate('touchStart', createTouchEventObject(50, 50));
  Card.simulate('touchMove', createTouchEventObject(250, 50));
  Card.simulate('touchEnd', createTouchEventObject(250, 50));
  
  expect(onSwipedRight).toHaveProperty('callCount', 1);
});

