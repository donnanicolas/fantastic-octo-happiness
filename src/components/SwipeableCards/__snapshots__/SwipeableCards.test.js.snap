// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a basic button 1`] = `
<div>
  <div>
    <div
      className="jCprYf"
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      onTransitionEnd={[Function]}
      style={
        Object {
          "opacity": 1,
          "transform": "rotate(0deg) translate3d(0%, 0, 0)",
        }
      }
    >
      <div
        className="ivEioO thumbnail"
        style={Object {}}
      >
        <img
          alt="An Album"
          src="https://i.scdn.co/image/81aa003fb60a031abe1f72253cf663cff201ad48"
        />
        <div
          className="caption"
        >
          <h3
            className="duAZJV"
          >
            Subwoofer Lullaby
          </h3>
          <p
            className="duAZJV"
          >
            C418
             - 
            An Album
          </p>
          <div>
            <button
              className="jEtlWH btn btn-primary btn-block"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Preview
              <span
                className="glyphicon glyphicon-play"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="jAITJt ivEioO thumbnail"
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0.75)",
        }
      }
    >
      <img
        alt="Any Which Way EP"
        src="https://i.scdn.co/image/1e732874b19b8853dccd29a0e68959af87c8c1d5"
      />
      <div
        className="caption"
      >
        <h3
          className="duAZJV"
        >
          Any Which Way - Martin Buttrich Red Mix
        </h3>
        <p
          className="duAZJV"
        >
          Stacey Pullen
           - 
          Any Which Way EP
        </p>
        <div>
          <button
            className="jEtlWH btn btn-primary btn-block"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Preview
            <span
              className="glyphicon glyphicon-play"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
